src/abi.md

```diff
@@ -1,21 +1,209 @@
+
+r[abi.compatibilty.zst]
+Two types, `T` and `U`, are *abi compatible* if both have size 0 and alignment 1.
+
+r[abi.compatibility.option]
+If `T` is a type listed in [layout.enum.option](https://doc.rust-lang.org/stable/core/option/index.html#representation), then given `S` is a type with size 0 and alignment 1, `T` is *abi compatible* with the types [`core::option::Option<T>`], [`core::result::Result<T,S>`], and [`core::result::Result<S,T>`].
+
+
+r[abi.compatibility.fn-ptr]
+An [`fn`-ptr type] `T` is compatible with an [`fn`-ptr type] `U` if `T` and `U` have *abi compatible* tags.
+
+r[abi.compatibility.extern-tag]
+Two [abi tags][abi tag] are *abi compatible* if:
+* They are the same string, or
+* One tag is `"X"`, and the other is `"X-unwind"`
+
+r[abi.compatibility.signature]
+Two function signatures are compatible if:
+* The [abi tags][abi tag] of both signatures are *abi compatible*,
+* They have the same number of parameters, excluding C-varargs,
+* Each parameter of both signatures, in order, are *abi compatible*, and
+* Either both signatures have C-varargs, or neither signature does.
+
+> [!NOTE]
+> A signature is compatible with itself.
+
+r[abi.compatibility.simd-abi]
+A type has *simd abi requirements* if:
+* It is a type declared with the standard-library repr-attrbute `simd`,
+* It is a aggregate type[^1], which has a type with *simd abi requirements* as a field.
+
+> [!NOTE]
+> The `repr(simd)` attribute cannot be used by Rust code, only by the standard library.
+
+r[abi.compatibility.simd-target-feature]
+A type with *simd abi requirements* may have one or more [*salient target features*][target_feature] . In the case of an aggregate type, the set of [*salient target features*][target_feature] is the union of the set of [*salient target features*][target_feature] of each field with *simd abi requirements*.
+
+> [!TARGET-SPECIFIC]
+> On x86 and x86-64, the [*salient target features*][target_feature] of the `simd` types are:
+> * [`__m128`], [`__m128i`], [`__m128f`], and [`__m128d`] (128-bit vector types): `sse`
+> * [`__m256`], [`__m256i`], [`__m256f`], and [`__m256d`] (256-bit vector types): `avx`
+> * [`__m512`], [`__m512i`], [`__m512f`], and [`__m512d`] (512-bit vector types): `avx512f` and `avx512vl`
+
+r[abi.compatibility.call]
+A call to a function `f` via a function item or function pointer with a given signature `S` is valid if and only if the signature of the definition `f` is *compatible* with the signature `S`, and:
+* The ABI tag of the signature is `extern "Rust"`, or
+* If any parameter type, the return type, or the type of any argument passed via C-varargs has *simd abi requirements*, each [*salient target features*][target_feature] of that type is either set at both the definition site of the function, and at the call site, or is set at neither site.
+
+The behaviour a call that is not valid is undefined.
```

Comment:
Can you say where you got these rules from?

Reply:
From: https://doc.rust-lang.org/core/primitive.fn.html#abi-compatibility